clippy_lints/src/indexing_slicing.rs

```diff
@@ -0,0 +1,201 @@
+//! lint on indexing and slicing operations
+
+use crate::consts::{constant, Constant};
+use crate::utils::higher::Range;
+use crate::utils::{self, higher};
+use rustc::hir::*;
+use rustc::lint::*;
+use rustc::ty;
+use syntax::ast::RangeLimits;
+
+/// **What it does:** Checks for out of bounds array indexing with a constant
+/// index.
+///
+/// **Why is this bad?** This will always panic at runtime.
+///
+/// **Known problems:** Hopefully none.
+///
+/// **Example:**
+/// ```rust
+/// let x = [1,2,3,4];
+///
+/// // Bad
+/// x[9];
+/// &x[2..9];
+///
+/// // Good
+/// x[0];
+/// x[3];
+/// ```
+declare_clippy_lint! {
+    pub OUT_OF_BOUNDS_INDEXING,
+    correctness,
+    "out of bounds constant indexing"
+}
+
+/// **What it does:** Checks for usage of indexing or slicing. Does not report
+/// if we can tell that the indexing or slicing operations on an array are in
+/// bounds.
+///
+/// **Why is this bad?** Indexing and slicing can panic at runtime and there are
+/// safe alternatives.
+///
+/// **Known problems:** Hopefully none.
+///
+/// **Example:**
+/// ```rust
+/// let x = vec![0; 5];
+/// // Bad
+/// x[2];
+/// &x[2..100];
+/// &x[2..];
+/// &x[..100];
+///
+/// // Good
+/// x.get(2)
+/// x.get(2..100)
+/// x.get(2..)
+/// x.get(..100)
+/// ```
+declare_clippy_lint! {
+    pub INDEXING_SLICING,
+    restriction,
+    "indexing/slicing usage"
+}
+
+#[derive(Copy, Clone)]
+pub struct IndexingSlicingPass;
+
+impl LintPass for IndexingSlicingPass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(INDEXING_SLICING, OUT_OF_BOUNDS_INDEXING)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for IndexingSlicingPass {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        if let ExprIndex(ref a, ref b) = &expr.node {
```

Comment:
Why change the names of the variables from telling names `array` + `index` to just `a` + `b`?